### PR TITLE
Implement best-fit-with-coalescing

### PR DESCRIPTION
**Pull request by sonots**
Implementation of best-fit-with-coalescing algorithm for memory allocator.

In this way, we split free large sized allocated memory, and use it for a small sized block. The splitted block can be merged (coalesced) with non-used neighborhood free blocks to reconstruct a large sized free block.

## Summary of experiments on memory consumption

I compared `Before` and `After` by running examples under chainer.
This algorithm succeeded to reduce memory consumption well for NLP tasks whose input sizes are varying.

|           | Before           | After  | Ratio (After / Before) | Note                                            |
|-----------|------------------|--------|-----------------------|--------------------------------------------------|
| mnist     | 246MiB           | 246MiB | 1                     |                                                  |
| cifar     | 695MiB           | 664MiB | 0.95                  |                                                  |
| sentiment | 231MiB           | 233MiB | 1                     |                                                  |
| vae       | 163MiB           | 164MiB | 1                     |                                                  |
| word2vec  | 179MiB           | 162MiB | 0.94                  |                                                  |
| seq2seq   | 12GiB (maximum of the GPU) | 3GiB   | 0.25 (or less)               | Before: 20 times of `free_all_blocks` are called. After: 0 time |

### CODE REVIEW

**Comment by sonots** [comment]
I investigated the least alignment of cudaMalloc, and it was 512 bytes, not 256 bytes with GeForce GTX TITAN X and NVIDIA Tesla k80 (AWS p2.xlarge). 
See https://gist.github.com/sonots/41daaa6432b1c8b27ef782cd14064269

We should change the round_size from 256 bytes to 512 bytes.

**Comment by sonots** [comment]
I investigated the speed of cudaMalloc vs cuMemAlloc https://gist.github.com/sonots/7e11231af2f2cc6b1f519ad832632566. cuMemAlloc was pretty faster than cudaMalloc. Should we use cuMemAlloc?

**Comment by sonots** [comment]
I tried to train seq2seq example of https://github.com/chainer/chainer/pull/2070.

## Pre-Information

```
$ python seq2seq.py --epoch 1 --gpu 0
```

```
$ nvidia-smi
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.39                 Driver Version: 375.39                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 0000:02:00.0     Off |                  N/A |
| 22%   50C    P2   112W / 250W |   6007MiB / 12207MiB |     38%      Default |
+-------------------------------+----------------------+----------------------+
```

I applied a patch as below to see number of calls of `free_all_blocks`:

```diff
diff --git a/cupy/cuda/memory.pyx b/cupy/cuda/memory.pyx
index 49efdf0..8af6e09 100644
--- a/cupy/cuda/memory.pyx
+++ b/cupy/cuda/memory.pyx
@@ -384,6 +384,7 @@ cdef class SingleDeviceMemoryPool:
         free.append(mem)

     cpdef free_all_blocks(self):
+        print("free_all_blocks")
         self._free.clear()

     cpdef free_all_free(self):
```

## Before

```
epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  elapsed_time
1           5.7843                                                                     70.5747
```

I observed GPU memory growed, and `free_all_blocks()` were issued 20 times.

```
  1 free_all_blocks
  2 free_all_blocks
  3 free_all_blocks
  4 free_all_blocks
  5 free_all_blocks
  6 free_all_blocks
  7 free_all_blocks
  8 free_all_blocks
  9 free_all_blocks
 10 free_all_blocks
 11 free_all_blocks
 12 free_all_blocks
 13 free_all_blocks
 14 free_all_blocks
 15 free_all_blocks
 16 free_all_blocks
 17 free_all_blocks
 18 free_all_blocks
 19 free_all_blocks
 20 free_all_blocks
```

Below are some shots of nvidia-smi during running.

```
| 22%   36C    P2    90W / 250W |   2646MiB / 12207MiB |     35%      Default |
...
| 22%   41C    P2   118W / 250W |  10455MiB / 12207MiB |     33%      Default |
...
| 22%   46C    P2   131W / 250W |  10713MiB / 12207MiB |     43%      Default |
```

## After (that is, with this patch)

```
epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  elapsed_time
1           5.87647                                                                    53.7349
```

Although the aim of this PR is to reduce memory usage, the performance also became 1.3 times faster.

No `free_all_blocks` call was observed.

Memory usage growth was stopped at `3072MiB`

```
| 22%   62C    P2   140W / 250W |   3072MiB / 12207MiB |     54%      Default |
```

**Comment by sonots** [comment]
I compared by running examples under chainer.

|           | Before           | After  | Raio (After / Before) | Extra                                            |
|-----------|------------------|--------|-----------------------|--------------------------------------------------|
| mnist     | 246MiB           | 246Mib | 1                     |                                                  |
| cifar     | 695MiB           | 695Mib | 1                     |                                                  |
| sentiment | 231MiB           | 232MiB | 1                     |                                                  |
| vae       | 163MiB           | 164MiB | 1                     |                                                  |
| word2vec  | 179MiB           | 162MiB | 0.94                  |                                                  |
| seq2seq   | 12GiB (maximum of the GPU) | 3GiB   | 0.25                  | Before: 20 times of `free_all_blocks` are called. After: 0 time |

**Comment by sonots** [comment]
@okuta Removed `[WIP]` label. Could you review?

**Comment by okuta** [comment]
OK, I will.

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -332,32 +366,97 @@ cdef class PooledMemory(Memory):
...
+    cpdef void _grow_free_if_necessary(self, Py_ssize_t size):
+        """Extend bins (_free) size if necessary"""
+        current_size = len(self._free)
+        if current_size >= size:
+            return
+        growth_size = size - current_size
+        growth = [[] for i in range(growth_size)]
+        self._free.extend(growth)
+
+    cpdef list _split(self, Chunk chunk, Py_ssize_t size):
```
Comment (okuta) [inline cupy/cuda/memory.pyx]: How about using tuple?

File: cupy/cuda/memory.pyx [hunk 2]
```diff
@@ -396,21 +517,21 @@ cdef class SingleDeviceMemoryPool:
...
-        for v in six.itervalues(self._free):
+        for v in self._free:
             n += len(v)
         return n
 
     cpdef used_bytes(self):
         cdef Py_ssize_t size = 0
-        for mem in six.itervalues(self._in_use):
-            size += mem.size
+        for chunk in six.itervalues(self._in_use):
```
Comment (okuta) [inline cupy/cuda/memory.pyx]: How about using `self._in_use.itervalues()`?

File: cupy/cuda/memory.pyx (hunk 1 above)
Comment (sonots) [inline cupy/cuda/memory.pyx]: changed to use tuple.

File: cupy/cuda/memory.pyx (hunk 2 above)
Comment (sonots) [inline cupy/cuda/memory.pyx]: changed to do so.

**Comment by sonots** [comment]
Can I merge?

**Comment by okuta** [comment]
Please wait

**Comment by sonots** [comment]
11 doctest of chainer fails, but it seems not related with this branch (11 doctest fails with cupy master, too) ...

**Comment by okuta** [comment]
LGTM

**Pull request merged**